ddtrace/internal/writer/writer.py

```diff
@@ -420,14 +421,17 @@ def _flush_queue_with_client(self, client, raise_exc=False):
                 )
         finally:
             if config.health_metrics_enabled and self.dogstatsd:
+                # _metrics is updated when traces are queued.
+                # Here we get a deep copy of the nested metrics dictionary as early as possible
+                metrics = copy.deepcopy(self._metrics)
                 namespace = self.STATSD_NAMESPACE
                 # Note that we cannot use the batching functionality of dogstatsd because
                 # it's not thread-safe.
                 # https://github.com/DataDog/datadogpy/issues/439
                 # This really isn't ideal as now we're going to do a ton of socket calls.
                 self.dogstatsd.distribution("datadog.%s.http.sent.bytes" % namespace, len(encoded))
                 self.dogstatsd.distribution("datadog.%s.http.sent.traces" % namespace, n_traces)
```

Comment:
```suggestion
        finally:
              with self._conn_lck:
                    if config.health_metrics_enabled and self.dogstatsd:
                        # _metrics is updated when traces are queued.
                        namespace = self.STATSD_NAMESPACE
                        # Note that we cannot use the batching functionality of dogstatsd because
                        # it's not thread-safe.
                        # https://github.com/DataDog/datadogpy/issues/439
                        # This really isn't ideal as now we're going to do a ton of socket calls.
                        self.dogstatsd.distribution("datadog.%s.http.sent.bytes" % namespace, len(encoded))
                        self.dogstatsd.distribution("datadog.%s.http.sent.traces" % namespace, n_traces)
```